shell/platform/android/android_image_generator.cc

```diff
@@ -0,0 +1,193 @@
+      break;
+    case kPremul_SkAlphaType:
+      break;
+    default:
+      return false;
+  }
+
+  // Since this decoder is only installed for API 28 and 29, there is no
+  // supported way to work around this unfortunate copy. It's possible to
+  // produce an `SkImage` direcly from an `AHardwareBuffer`, but
+  // `AndroidBitmap_getHardwareBuffer` and `Bitmap.getHardwareBuffer` are only
+  // available in API 30+ and 31+ respectively.
+  memcpy(pixels, decoded_data_->data(), decoded_data_->size());
+  return true;
+}
+
+void AndroidImageGenerator::DecodeImage() {
+  DoDecodeImage();
+
+  header_decoded_latch_.Signal();
+  fully_decoded_latch_.Signal();
+}
+
+void AndroidImageGenerator::DoDecodeImage() {
+  if (!g_flutter_jni_class || !g_decode_image_method) {
+    return;
+  }
+
+  // Call FlutterJNI.decodeImage
+
+  JNIEnv* env = fml::jni::AttachCurrentThread();
+
+  fml::jni::ScopedJavaLocalRef<jobject> direct_buffer(
+      env, env->NewDirectByteBuffer(const_cast<void*>(data_->data()),
+                                    data_->size()));
+  fml::jni::ScopedJavaGlobalRef<jobject>* bitmap =
+      new fml::jni::ScopedJavaGlobalRef(
+          env, env->CallStaticObjectMethod(g_flutter_jni_class->obj(),
+                                           g_decode_image_method,
+                                           direct_buffer.obj(), this));
+  FML_CHECK(fml::jni::CheckException(env));
+
+  AndroidBitmapInfo info;
+  int status;
+  if ((status = AndroidBitmap_getInfo(env, bitmap->obj(), &info)) < 0) {
+    FML_DLOG(ERROR) << "Failed to get bitmap info, status=" << status;
+    return;
+  }
+  FML_DCHECK(info.format == ANDROID_BITMAP_FORMAT_RGBA_8888);
+
+  // Lock the android buffer in a shared pointer
+
+  void* pixel_lock;
+  if ((status = AndroidBitmap_lockPixels(env, bitmap->obj(), &pixel_lock)) <
+      0) {
+    FML_DLOG(ERROR) << "Failed to lock pixels, error=" << status;
+    return;
+  }
+
+  SkData::ReleaseProc on_release = [](const void* ptr, void* context) -> void {
+    fml::jni::ScopedJavaGlobalRef<jobject>* bitmap =
+        reinterpret_cast<fml::jni::ScopedJavaGlobalRef<jobject>*>(context);
+    auto env = fml::jni::AttachCurrentThread();
+    AndroidBitmap_unlockPixels(env, bitmap->obj());
+  };
+
+  decoded_data_ = SkData::MakeWithProc(
+      pixel_lock, info.width * info.height * sizeof(uint32_t), on_release,
+      bitmap);
+}
+
+bool AndroidImageGenerator::Register(JNIEnv* env) {
+  g_flutter_jni_class = new fml::jni::ScopedJavaGlobalRef<jclass>(
+      env, env->FindClass("io/flutter/embedding/engine/FlutterJNI"));
+  if (g_flutter_jni_class->is_null()) {
+    FML_LOG(ERROR) << "Failed to find FlutterJNI class";
+    return false;
+  }
+
+  g_decode_image_method = env->GetStaticMethodID(
+      g_flutter_jni_class->obj(), "decodeImage",
+      "(Ljava/nio/ByteBuffer;J)Landroid/graphics/Bitmap;");
+
+  if (g_decode_image_method == nullptr) {
+    FML_LOG(ERROR) << "Failed to find FlutterJNI.decodeImage method";
```

Comment:
We should only log here if there's some user action to take. This strikes me as a "this should never happen" thing, and we should probably just DCHECK it and have a test that covers this path.

Reply:
Fixed.